relayer/pkg/chainlink/txm/txm.go

```diff
@@ -69,135 +81,167 @@ func (txm *starktxm) Name() string {
+			hash, err := txm.broadcast(ctx, privKey, tx.accountAddress, tx.call)
+			if err != nil {
+				txm.lggr.Errorw("transaction failed to broadcast", "error", err, "tx", tx.call)
+			} else {
+				txm.lggr.Infow("transaction broadcast", "txhash", hash)
 			}
-			wg.Wait()
-			tick = time.After(utils.WithJitter(txm.cfg.TxSendFrequency()) - time.Since(start))
-		case <-txm.stop:
-			return
 		}
 	}
 }
 
 const FEE_MARGIN uint64 = 115
 
-func (txm *starktxm) broadcastBatch(ctx context.Context, privKey string, accountAddress caigotypes.Hash, txs []caigotypes.FunctionCall) (txhash string, err error) {
+func (txm *starktxm) broadcast(ctx context.Context, privKey string, accountAddress caigotypes.Hash, tx caigotypes.FunctionCall) (txhash string, err error) {
+	txs := []caigotypes.FunctionCall{tx}
+	client, err := txm.client.Get()
+	if err != nil {
+		return txhash, fmt.Errorf("broadcast: failed to fetch client: %+w", err)
+	}
 	// create new account
-	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), txm.client.Gw, caigo.AccountVersion1)
+	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), client.Gw, caigo.AccountVersion1)
+	if err != nil {
+		return txhash, fmt.Errorf("failed to create new account: %+w", err)
+	}
+
+	nonce, err := txm.nonce.NextSequence(accountAddress, client.Gw.ChainId)
 	if err != nil {
-		return txhash, errors.Errorf("failed to create new account: %s", err)
+		return txhash, fmt.Errorf("failed to get nonce: %+w", err)
 	}
 
 	// get fee for txm
+	// optional - pass nonce to fee estimate (if nonce gets ahead, estimate may fail)
+	// can we estimate fee without calling estimate - tbd with 1.0
 	feeEstimate, err := account.EstimateFee(ctx, txs, caigotypes.ExecuteDetails{})
 	if err != nil {
-		return txhash, errors.Errorf("failed to estimate fee: %s", err)
+		return txhash, fmt.Errorf("failed to estimate fee: %+w", err)
 	}
 
 	fee, _ := big.NewInt(0).SetString(string(feeEstimate.OverallFee), 0)
 	expandedFee := big.NewInt(0).Mul(fee, big.NewInt(int64(FEE_MARGIN)))
 	max := big.NewInt(0).Div(expandedFee, big.NewInt(100))
 	details := caigotypes.ExecuteDetails{
 		MaxFee: max,
+		Nonce:  nonce,
 	}
 
-	// TODO: investigate if nonce management is needed (nonce is requested queried by the sdk for now)
 	// transmit txs
 	execCtx, execCancel := context.WithTimeout(ctx, txm.cfg.TxTimeout()*time.Second)
 	defer execCancel()
 	res, err := account.Execute(execCtx, txs, details)
 	if err != nil {
-		return txhash, errors.Errorf("failed to invoke tx: %s", err)
+		// TODO: handle initial broadcast errors - what kind of errors occur?
+		return txhash, fmt.Errorf("failed to invoke tx: %+w", err)
 	}
 
 	// handle nil pointer
 	if res == nil {
 		return txhash, errors.New("execute response and error are nil")
 	}
 
-	return res.TransactionHash, nil
+	// update nonce if transaction is successful
+	err = errors.Join(
+		txm.nonce.IncrementNextSequence(accountAddress, client.Gw.ChainId, nonce),
```

Comment:
Since this is technically fallible, why not do the increment optimistically? e.g. `txm.nonce.NextSequence()` should return the next nonce to use, and increase the counter by 1. That way it's an atomic operation -- if transactions fail we just end up using the next nonce in the sequence after that

Reply:
afaik it's more robust to split it into a get + increment functionality

for example: if the tx encounters an error (estimation failure, signing issues, RPC rejecting, client errors), you wouldn't want to increment the tx until you are sure that the tx was accepted by the RPC. otherwise it would be very easy to fall out of sync

Reply:
do the transactions need to be sequential? If not it's perfectly time if the id doesn't end up being used because of an error etc

Reply:
based on this it sounds like they need to be sequential: https://docs.starknet.io/documentation/architecture_and_concepts/State/starknet-state/

> Similarly to Ethereum, the contract’s nonce is used to provide replay protection at the protocol level. When a transaction is sent to a given contract, the transaction’s nonce must match the nonce of the contract. After the transaction is executed, the contract’s nonce is increased by one.